EIPS/eip-draft-whisper.md

```diff
@@ -0,0 +1,138 @@
+
+If Whisper node does not support a particular packet code, it should just ignore the packet without generating any error.
+
+### Packet Codes
+
+The Whisper sub-protocol should support the following packet codes:
+
+| EIP   | Name                       | Int Value |
+|-------|----------------------------|-----------|
+|       | Status                     |   0x00    |
+|       | Messages                   |   0x01    |
+|       | P2P                        |   0x02    |
+|       | P2P Request                |   0x03    |
+|       | PoW Requirement            |   0x04    |
+|       | Bloom Filter               |   0x05    |
+
+
+### Packet Format and Usage
+
+**Status** [`0x00`, `whisper_protocol_version`] 
+
+This packet contains two objects: integer (0x00) followed by integer (Whisper protocol version).
+
+Informs a peer of the Whisper version. This message should be send after the initial handshake and prior to any other messages.
+
+**Messages** [`0x01`, `whisper_envelope`]
+
+This packet contains two objects: integer (0x01) followed by a single Whisper Envelope.
+
+This packet is used for sending the standard Whisper envelopes.
+
+**P2P** [`0x02`, `whisper_envelope`]
+
+This packet contains two objects: integer (0x02) followed by a single Whisper Envelope.
+
+This packet is used for sending the peer-to-peer messages, which are not supposed to be forwarded any further. E.g. it might be used by the Whisper Mail Server for delivery of old (expired) messages, which is otherwise not allowed.
+
+**P2P Request** [`0x03`, `whisper_envelope`]
+
+This packet contains two objects: integer (0x03) followed by a single Whisper Envelope.
+
+This packet is used for sending Dapp-level peer-to-peer requests, e.g. Whisper Mail Client requesting old messages from the Whisper Mail Server.
+
+**PoW Requirement** [`0x04`, `PoW`]
+
+This packet contains two objects: integer (0x04) followed by a single floating point value of PoW.
+
+This packet is used by Whisper nodes for dynamic adjustment of their individual PoW requirements. Receipient of this message should no longer deliver the sender messages with PoW lower than specified in this message.
+
+**Bloom Filter** [`0x05`, `bytes`]
+
+This packet contains two objects: integer (0x05) followed by a byte array of arbitrary size.
+
+This packet is used by Whisper nodes for sharing their interest in messages with specific topics.
+
+### Whisper Envelope
+
+Envelopes are RLP-encoded structures of the following format:
+
+	[ Version, Expiry, TTL, Topic, AESNonce, Data, EnvNonce ]
+	
+`Version`: up to 4 bytes (currently one byte containing zero). Version indicates encryption method. If Version is higher than current, envelope could not be decoded, and therefore only forwarded to the peers.
+
+`Expiry`: 4 bytes (UNIX time in seconds).
+
+`TTL`: 4 bytes (time-to-live in seconds).
+
+`Topic`: 4 bytes of arbitrary data.
```

Comment:
You can have other topics though. For something like a broadcast you may multicast to multiple topics rather than only to a single one. This opens up the ability to use one-time keys for convenient broadcast with a pretty simple scheme (using a chat room as a concrete example):
  - Encrypt the message with a 32-byte one-time use key K (and globally agreed-upon nonce). Call the ciphertext C
  - Topics: [sha3<sup>2</sup>(chat_room_name_1), sha3<sup>2</sup>(chat_room_name_2), ...]
  - Payload: [sha3(chat_room_name_1) ^ K, sha3(chat_room_name_2) ^ K, ..., C] 

The payload length is always (32 * n_topics) + len(C).
Now anybody who knows the chat room names can decrypt the messages by 
  a) checking for the topic
  b) extracting the one-time key
  c) decrypting the plaintext